From 745b53fa6f00ffcf65ee1ff3f24d0bd26dec2594 Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Fri, 18 Oct 2024 14:25:18 +0800
Subject: [PATCH 64/68] MMFMWK-9383 waylandsink: fix display abnormal on Xen

On Xen, sw decoder uses waylandsink proposed buffer pool that width
and height are 16 aligned, but waylandsink cannot create wlbuffer
using dmabuf.

If waylandsink received buffer has videometa, uses it to update
videoinfo and create shm wlbuffer. Otherwise, copy into internal
buffer pool.

Upstream-Status: Pending
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 ext/wayland/gstwaylandsink.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index aac4ea8..1ef5247 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -1271,15 +1271,30 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
   }
 
 handle_shm:
+  GstVideoInfo old_vinfo = self->video_info;
+  GstVideoMeta *vmeta = gst_buffer_get_video_meta (buffer);
+  if (vmeta) {
+      for (gint i = 0; i < vmeta->n_planes; i++) {
+        self->video_info.offset[i] = vmeta->offset[i];
+        self->video_info.stride[i] = vmeta->stride[i];
+      }
+      self->video_info.width = vmeta->width + vmeta->alignment.padding_left +
+          vmeta->alignment.padding_right;
+      self->video_info.height = vmeta->height + vmeta->alignment.padding_bottom +
+          vmeta->alignment.padding_top;
+      self->video_info.size = gst_buffer_get_size (buffer);
+  }
+
   if (!wbuf && gst_wl_display_check_format_for_shm (self->display,
           &self->video_info)) {
-    if (gst_buffer_n_memory (buffer) == 1 && gst_is_fd_memory (mem))
+    if (vmeta && gst_buffer_n_memory (buffer) == 1 && gst_is_fd_memory (mem))
       wbuf = gst_wl_shm_memory_construct_wl_buffer (mem, self->display,
           &self->video_info);
 
     /* If nothing worked, copy into our internal pool */
     if (!wbuf) {
       GstVideoFrame src, dst;
+      self->video_info = old_vinfo;
 
       /* we don't know how to create a wl_buffer directly from the provided
        * memory, so we have to copy the data to shm memory that we know how
-- 
2.43.0

